c/src/lib/libbsp/arm/shared/stm32fxxxx/sdram/hal-sdram.c

```diff
@@ -0,0 +1,283 @@
+  SDRAM_Timing.SelfRefreshTime      = 4;
+  SDRAM_Timing.RowCycleDelay        = 7;
+  SDRAM_Timing.WriteRecoveryTime    = 2;
+  SDRAM_Timing.RPDelay              = 2;
+  SDRAM_Timing.RCDDelay             = 2;
+
+
+  hsdram.Init.SDBank             = FMC_SDRAM_BANK1;
+  hsdram.Init.ColumnBitsNumber   = FMC_SDRAM_COLUMN_BITS_NUM_8;
+  hsdram.Init.RowBitsNumber      = FMC_SDRAM_ROW_BITS_NUM_12;
+  hsdram.Init.MemoryDataWidth    = SDRAM_MEMORY_WIDTH;
+  hsdram.Init.InternalBankNumber = FMC_SDRAM_INTERN_BANKS_NUM_4;
+  hsdram.Init.CASLatency         = FMC_SDRAM_CAS_LATENCY_2;
+  hsdram.Init.WriteProtection    = FMC_SDRAM_WRITE_PROTECTION_DISABLE;
+  hsdram.Init.SDClockPeriod      = SDCLOCK_PERIOD;
+  hsdram.Init.ReadBurst          = FMC_SDRAM_RBURST_ENABLE;
+  hsdram.Init.ReadPipeDelay      = FMC_SDRAM_RPIPE_DELAY_0;
+
+  /* Initialize the SDRAM controller */
+  if(HAL_SDRAM_Init(&hsdram, &SDRAM_Timing) != HAL_OK)
+  {
+    /* Initialization Error */
+    stm32f_error_handler();
+  }
+
+  /* Program the SDRAM external device */
+  BSP_SDRAM_Initialization_Sequence(&hsdram, &command);
+}
+
+/**
+ * @brief SDRAM MSP Initialization
+ *        This function configures the hardware resources used in this example:
+ *           - Peripheral's clock enable
+ *           - Peripheral's GPIO Configuration
+ * @param hsram: SDRAM handle pointer
+ * @retval None
+ */
+void HAL_SDRAM_MspInit(
+  SDRAM_HandleTypeDef *hsdram
+)
+{
+  GPIO_InitTypeDef GPIO_Init_Structure;
+
+  /*##-1- Enable peripherals and GPIO Clocks #################################*/
+  /* Enable GPIO clocks */
+  __HAL_RCC_GPIOC_CLK_ENABLE();
+  __HAL_RCC_GPIOD_CLK_ENABLE();
+  __HAL_RCC_GPIOE_CLK_ENABLE();
+  __HAL_RCC_GPIOF_CLK_ENABLE();
+  __HAL_RCC_GPIOG_CLK_ENABLE();
+  __HAL_RCC_GPIOH_CLK_ENABLE();
+
+  /* Enable FMC clock */
+  __HAL_RCC_FMC_CLK_ENABLE();
+
+  /*##-2- Configure peripheral GPIO ##########################################*/
+  GPIO_Init_Structure.Mode = GPIO_MODE_AF_PP;
+  GPIO_Init_Structure.Pull = GPIO_PULLUP;
+  GPIO_Init_Structure.Speed = GPIO_SPEED_FAST;
+  GPIO_Init_Structure.Alternate = GPIO_AF12_FMC;
+
+  /* GPIOC configuration */
+  GPIO_Init_Structure.Pin = GPIO_PIN_3;
+  HAL_GPIO_Init(GPIOC, &GPIO_Init_Structure);
+
+  /* GPIOD configuration */
+  GPIO_Init_Structure.Pin = GPIO_PIN_0 | GPIO_PIN_1 | GPIO_PIN_3 | GPIO_PIN_8 | GPIO_PIN_9 |
+  GPIO_PIN_10 | GPIO_PIN_14 | GPIO_PIN_15;
+  HAL_GPIO_Init(GPIOD, &GPIO_Init_Structure);
+
+  /* GPIOE configuration */
+  GPIO_Init_Structure.Pin = GPIO_PIN_0 | GPIO_PIN_1 | GPIO_PIN_7| GPIO_PIN_8 | GPIO_PIN_9 |\
+ GPIO_PIN_10 | GPIO_PIN_11 | GPIO_PIN_12 | GPIO_PIN_13 | GPIO_PIN_14 |\
+ GPIO_PIN_15;
+  HAL_GPIO_Init(GPIOE, &GPIO_Init_Structure);
+
+  /* GPIOF configuration */
+  GPIO_Init_Structure.Pin = GPIO_PIN_0 | GPIO_PIN_1 | GPIO_PIN_2| GPIO_PIN_3 | GPIO_PIN_4 |\
+ GPIO_PIN_5 | GPIO_PIN_11 | GPIO_PIN_12 | GPIO_PIN_13 | GPIO_PIN_14 |\
+ GPIO_PIN_15;
+  HAL_GPIO_Init(GPIOF, &GPIO_Init_Structure);
```

Comment:
This is boiler plate initialization code for the the GPIO pins used for an external SDRAM bus.  It comes from an ST example project.